x/datapool/keeper/pool.go

```diff
@@ -328,3 +322,121 @@ func (k Keeper) DeployAndRegisterNFTContract(ctx sdk.Context, wasmCode []byte) (
 
 	return contractAddr, nil
 }
+
+func (k Keeper) BuyDataAccessNFT(ctx sdk.Context, buyer sdk.AccAddress, poolID, round uint64, payment sdk.Coin) error {
+	pool, err := k.GetPool(ctx, poolID)
+	if err != nil {
+		return sdkerrors.Wrapf(err, "failed to get pool %d", poolID)
+	}
+
+	if pool.GetNumIssuedNfts() == pool.GetPoolParams().GetMaxNftSupply() {
+		return types.ErrNFTAllIssued
+	}
+
+	if pool.GetRound() != round {
+		return types.ErrRoundNotMatched
+	}
+
+	if !payment.Equal(pool.GetPoolParams().GetNftPrice()) {
+		return types.ErrPaymentNotMatched
+	}
+
+	poolAcc, err := sdk.AccAddressFromBech32(pool.GetPoolAddress())
+	if err != nil {
+		return sdkerrors.Wrapf(err, "invalid pool address")
+	}
+
+	err = k.bankKeeper.SendCoins(ctx, buyer, poolAcc, sdk.NewCoins(payment))
+	if err != nil {
+		return sdkerrors.Wrapf(err, "failed to pay")
+	}
+
+	//mint data access NFT when pool is activated
+	if pool.GetStatus() == types.ACTIVE {
+		contractAddr := pool.GetNftContractAddr()
+
+		contractAcc, err := sdk.AccAddressFromBech32(contractAddr)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "invalid NFT contract address")
+		}
+
+		mintMsg := types.NewMsgMintDataAccessNFT(pool.GetNumIssuedNfts()+1, buyer.String())
+		mintMsgBz, err := json.Marshal(mintMsg)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "failed to marshal mint NFT Msg")
+		}
+
+		_, err = k.wasmKeeper.Execute(ctx, contractAcc, poolAcc, mintMsgBz, sdk.NewCoins(types.ZeroFund))
+		if err != nil {
+			return sdkerrors.Wrapf(err, "failed to mint NFT")
+		}
+	} else {
+		// if pool is PENDING state, add buyer to white list
+		k.AddToWhiteList(ctx, poolID, buyer)
+	}
```

Comment:
Minting NFT will be modified after we make a decision of how and when to distribute NFTs.

Reply:
Right.
In my opinion, when the state of the Pool becomes ACTIVE, NFTs should be distributed along with ShareToken.

Reply:
Agreed. I'm gonna modify this part to distribute NFTs as what you've done in #311